src/kvstore/collectives/src/collectives.cc

```
@@ -0,0 +1,855 @@
+  std::string ops_broadcast   = OPS_BROADCAST;
+  std::string new_key;
+  size_t idx = 0;
+  for (auto& key : keys) {
+    // To simplify original logic for group key value, we rename the original
+    // duplicated key and make every key unique now.
+    size_t index = countIDX(keys, key, idx);
+    new_key = ops_prefix + delimiter + ops_broadcast + delimiter +
+              key_prefix + delimiter + std::to_string(key) + delimiter +
+              idx_prefix + delimiter + std::to_string(index);
+    v_keys.push_back(new_key);
+    idx++;
+  }
+  return MXBroadcastImpl(v_keys, values, root_rank, priority);
+}
+
+int MXBroadcastEx(const std::vector<std::string> &keys,
+                  const std::vector<mxnet::NDArray*> &values,
+                  int root_rank,
+                  int priority) {
+  std::vector<std::string> v_keys;
+  std::string key_prefix  = STR_PREFIX;
+  std::string idx_prefix  = IDX_PREFIX;
+  std::string delimiter   = DELIMITER;
+  std::string ops_prefix  = OPS_PREFIX;
+  std::string ops_broadcast   = OPS_BROADCAST;
+  std::string new_key;
+  size_t idx = 0;
+  for (auto& key : keys) {
+    // To simplify original logic for group key value pairs, we rename the original
+    // duplicated key and make every key unique now.
+    size_t index = countIDX(keys, key, idx);
+    new_key = ops_prefix + delimiter + ops_broadcast + delimiter +
+              key_prefix + delimiter + key + delimiter +
+              idx_prefix + delimiter + std::to_string(index);
+    v_keys.push_back(new_key);
+    idx++;
+  }
+  return MXBroadcastImpl(v_keys, values, root_rank, priority);
+}
+
+int MXAllGather(const std::vector<int> &keys,
+                const std::vector<mxnet::NDArray*> &values,
+                int priority) {
+  // place holder
+  return 0;
+}
+
+int MXAllGatherEx(const std::vector<std::string> &keys,
+                  const std::vector<mxnet::NDArray*> &values,
+                  int priority) {
+  // place holder
```

If it's not implemented, could you raise an error instead of returning 0 silently?

---

OK.